### PR TITLE
feat(arrow/array): add ValueAsAny for native Go values

### LINKED ISSUE #462: Add method to Array interface to retrieve native Go values
### Describe the enhancement requested

In downstream projects using the Apache Arrow Go library — particularly in driver implementations and generic data ingestion layers — there’s often a need to extract values from an arrow.Array without knowing its concrete type ahead of time.

Currently, this requires verbose type switches and manual downcasting for each array type, which adds boilerplate and makes generic handling of data more complex.

To simplify this pattern, we propose adding a method to the Array interface that returns the Go-native equivalent of the value at a given index. For example:

ValueAsAny(index int) any

This change would significantly reduce boilerplate and improve ergonomics in downstream projects that need to interface with Arrow arrays in a type-agnostic way.

### Component(s)

Other

### PR DESCRIPTION
**Pull request by dgvj-work**
### Rationale for this change

Downstream code often needs type-agnostic access to Arrow values as native Go types. `GetOneForMarshal` is JSON-oriented (e.g. `int8` → `float64`, timestamps/decimals as strings), which forces boilerplate type switches for generic ingestion.

Fixes #462

### What changes are included in this PR?

- Add `array.ValueAsAny(arr, i)` plus an optional `ValueAsAnyer` interface, without adding a method to `arrow.Array`
- Implement it across array types and common extensions so nulls return `nil` and values stay native (`int8`, `arrow.Timestamp`, decimal nums, `[]any` for lists, ordered `[name, value]` pairs for structs)
- Union slots keep their type id when the selected child is null (`[]any{typeID, nil}`)
- Variant decode errors are returned as the `any` value rather than collapsed to `nil`
- Add unit tests covering primitives, temporal/decimal, nested, dictionary, union, duplicate struct fields, and variant decode errors, including differences vs `GetOneForMarshal`

### Are these changes tested?

- `go test ./arrow/ ./arrow/array/ ./arrow/extensions/`

### Are there any user-facing changes?

Yes — new `array.ValueAsAny` helper for native Go value access.

### CODE REVIEW

File: arrow/array.go [hunk 1]
```diff
@@ -110,6 +110,12 @@ type Array interface {
 	// ValueStr returns the value at index as a string.
 	ValueStr(i int) string
 
+	// ValueAsAny returns the native Go value at index i, or nil if the slot is null.
+	// Unlike GetOneForMarshal, values are not converted for JSON encoding
+	// (for example int8 stays int8, timestamps stay arrow.Timestamp, and
+	// nested values are []any / map[string]any of native values).
+	ValueAsAny(i int) any
```
Comment (zeroshade) [inline arrow/array.go]: Adding a method to the exported `arrow.Array` interface is source-incompatible for downstream implementations. A type satisfying the interface before this commit now fails with `missing method ValueAsAny`.

Unless this targets an explicitly breaking release, could this be exposed through a standalone helper or narrower optional interface instead?

File: arrow/array/union.go [hunk 2]
```diff
@@ -333,6 +333,19 @@ func (a *SparseUnion) GetOneForMarshal(i int) interface{} {
 	return []interface{}{typeID, data.GetOneForMarshal(i)}
 }
 
+func (a *SparseUnion) ValueAsAny(i int) any {
+	typeID := a.RawTypeCodes()[i]
+
+	childID := a.ChildID(i)
+	data := a.Field(childID)
+
+	if data.IsNull(i) {
```
Comment (zeroshade) [inline arrow/array/union.go]: This union slot itself is valid and still has a meaningful type code even when its selected child is null. Returning bare `nil` loses that discriminator.

Please preserve the union representation, for example `[]any{typeID, nil}`, consistently with non-null union values.

File: arrow/array/struct.go [hunk 3]
```diff
@@ -280,6 +280,18 @@ func (a *Struct) GetOneForMarshal(i int) interface{} {
...
 }
 
+func (a *Struct) ValueAsAny(i int) any {
+	if a.IsNull(i) {
+		return nil
+	}
+	tmp := make(map[string]any)
+	fieldList := a.data.dtype.(*arrow.StructType).Fields()
+	for j, d := range a.fields {
+		tmp[fieldList[j].Name] = d.ValueAsAny(i)
```
Comment (zeroshade) [inline arrow/array/struct.go]: Arrow permits duplicate struct field names, but assigning by name into a map silently overwrites earlier fields. For `dup:int32=11, dup:int64=22`, this returns only `{"dup": 22}`.

Please use a representation that preserves field order and duplicates, or explicitly reject ambiguous schemas.

File: arrow/extensions/variant.go [hunk 4]
```diff
@@ -612,6 +612,17 @@ func (v *VariantArray) GetOneForMarshal(i int) any {
 	return val.Value()
 }
 
+func (v *VariantArray) ValueAsAny(i int) any {
+	if v.IsNull(i) {
+		return nil
+	}
+	val, err := v.Value(i)
+	if err != nil {
```
Comment (zeroshade) [inline arrow/extensions/variant.go:620]: A decode error on a non-null variant is converted to `nil`, making malformed data indistinguishable from an actual null slot. The reproduced case has `IsNull(i) == false` and invalid metadata, but `ValueAsAny(i) == nil`.

Please preserve an error signal rather than silently converting corruption to null.

**Review by zeroshade — CHANGES_REQUESTED** [review]
Found four public-API correctness and compatibility issues that need to be addressed before this can merge; see inline comments on `arrow/array.go`, `arrow/array/union.go`, `arrow/array/struct.go`, and `arrow/extensions/variant.go`.

The remaining implementations look consistent across primitive, temporal, decimal, nested, dictionary, encoded, and extension arrays. Exact-head package, race, repeated, and 32-bit tests passed.

---

> *This review was drafted by an AI-assisted tool and confirmed by an Arrow Go maintainer. After you've addressed the points above and pushed an update, an Arrow Go maintainer — a real person — will take the next look at the PR. The findings cite the project's review criteria; if you think one of them is mis-applied, please reply on the PR and a maintainer will weigh in.*
>
> *More on how Arrow Go handles maintainer review:* [CONTRIBUTING.md](https://github.com/apache/arrow-go/blob/main/CONTRIBUTING.md).

File: arrow/array.go (hunk 1 above)
Comment (dgvj-work) [reply arrow/array.go]: Moved it off `arrow.Array`. Callers can use `array.ValueAsAny(arr, i)`, and custom array types can opt in via the optional `ValueAsAnyer` interface.

File: arrow/array/union.go (hunk 2 above)
Comment (dgvj-work) [reply arrow/array/union.go]: Updated. A null child now returns `[]any{typeID, nil}` so the union type id is preserved.

File: arrow/extensions/variant.go (hunk 4 above)
Comment (dgvj-work) [reply arrow/extensions/variant.go:620]: Decode errors are now returned as the `any` value (the `error` itself) instead of `nil`, so a corrupt non-null slot is distinguishable from an actual null.

File: arrow/array/struct.go (hunk 3 above)
Comment (dgvj-work) [reply arrow/array/struct.go]: Structs now return ordered `[name, value]` pairs instead of a map, so field order and duplicate names are preserved.

**Comment by dgvj-work** [comment]
Hi @zeroshade - I've addressed all four review points in the latest commit (6293617) and replied on each thread. Ready for another look when you have a moment - and for CI approval once you're happy with the changes. Thanks!

**Review by zeroshade — APPROVED** [review]
Thanks for the updates! All four concerns are addressed — the standalone `array.ValueAsAny` + optional `ValueAsAnyer` interface avoids the breaking change to `arrow.Array`, unions preserve the type code for null children, structs preserve field order and duplicate names, and variant decode errors are no longer silently converted to null.

One note (non-blocking): returning the `error` itself as the `any` value from `VariantArray.ValueAsAny` works, but callers need to type-assert to distinguish it from a legitimate value. Fine for now given the single-return signature.

---
*This review was drafted with the help of an AI-assisted tool and may contain mistakes; it was reviewed and confirmed by an Apache Arrow maintainer before posting. See the [contributing guide](https://github.com/apache/arrow-go/blob/main/CONTRIBUTING.md) for what the project considers a maintainer review.*
